### ISSUE TITLE
Contrast is significantly different in current version

### ISSUE THREAD

**Issue by bbqsrc** [matklad/pale-fire#10]
Hey! My Pale Fire theme just updated automatically in VS Code, and up until now I have been very happy with the theme.

However, the changes to the colour contrast on the UI elements is making it difficult for me to distinguish errors from warnings on the scroll bar, and file names in the explorer bar.

I am slightly colour blind, so these two colours look almost exactly the same to me unless I focus really hard:

<img width="16" alt="image" src="https://user-images.githubusercontent.com/279099/87540163-7ac43480-c69f-11ea-8288-a56b45303e89.png">

Would it be possible to either raise the contrast of those elements back to their previous levels, or, alternatively, offer a "High Contrast" variant in the same extension for my bad eyes? :smile:

**Comment by matklad** [comment]
Note that in the mean time you can revert to an older version.

Another nice alternative is https://github.com/drzix/hc-zenburn-vscode

But I do think that for these colors it makes sense to increase the contrast

**Comment by bbqsrc** [comment]
Oh! I never knew you could downgrade extensions in VS Code. Thanks!

**Comment by arzg** [comment]
Sorry about that, [I was the one that made those changes](https://github.com/matklad/pale-fire/pull/4).

I think there are a quite few places in the UI where that red has been used, and I don’t think that [the Zenburn palette](https://github.com/matklad/pale-fire/blob/59fca48c1eb1dd17494594c19d556c0a5b5ba1be/themes/Pale%20Fire-color-theme.json#L390) offers any higher contrast colours.

Perhaps it makes sense to deviate from the palette and introduce more saturated versions of the red (and possibly other colours) just for the UI.

@bbqsrc Are there any other parts of the UI you find are difficult to see and/or distinguish? I want to make sure that my legacy isn’t making colour schemes so unreadable that people downgrade 😅

**Comment by bbqsrc** [comment]
The contrast is rather poor for my eyes in the file explorer, compare before:

<img width="291" alt="image" src="https://user-images.githubusercontent.com/279099/87541505-a2b49780-c6a1-11ea-8109-989216c2323a.png">

After:

<img width="292" alt="image" src="https://user-images.githubusercontent.com/279099/87541522-a9430f00-c6a1-11ea-9171-e90eed083445.png">

The change in the lightness of the background, and then also the lightening of the text, has made my eyes have to strain more to read the text.

----

It has affected the contrast in the terminal as well:


<img width="417" alt="image" src="https://user-images.githubusercontent.com/279099/87541702-f7581280-c6a1-11ea-8354-4b88e887ca2c.png">

vs

<img width="426" alt="image" src="https://user-images.githubusercontent.com/279099/87541653-dc859e00-c6a1-11ea-8717-4e4be57d3ab0.png">

Everything becoming muted there is perhaps not ideal for me.

It feels like applying a absolute zenburn palette to the non-text editing area with its rather muted affect is not ideal for me, and what attracted to me to pale fire in the first place was that it was a much less muted version of zenburn with the right mix of colors that my brain can coordinate what code means much quicker than with other themes.

This is a tool I have used to check if a set of colours meets WCAG 2.0 AA compliance, the accessibility standard for the web: https://webaim.org/resources/contrastchecker/ -- maybe this will help with making colour decisions in the future? :)

But please don't take any of my criticisms above to heart, please keep iterating at your own pace, I'm just very happy to have these discussions and try to not be an unreasonable demander of other people's time. :smile:

**Comment by arzg** [comment]
> Everything becoming muted there is perhaps not ideal for me.
>
> It feels like applying a absolute zenburn palette to the non-text editing area with its rather muted affect is not ideal for me

I personally find it easy to read, but this is where themes become very subjective. I would prefer to avoid a situation in which anyone finds the theme so difficult to read they have to stick to an older version, so I think there are two choices:

- Add additional, higher-contrast colours to the palette and use them in the UI
- Add an option to revert to the old UI

The first adds more colours to the palette (which is already quite long, although I guess colours can always be removed), but brings with it the risk of losing the ‘alien fruit salad-ness’ that the Zenburn palette has.

The second adds complexity. In my experience, themes that want to provide options tend to do so by providing multiple themes that the user can switch between. Take the default Dark and Dark+ themes: identical, except from Dark+ highlighting many more scopes. The duplication between the themes is removed through [the use of includes](https://github.com/microsoft/vscode/blob/91a3e6b0db4c0b39eddc38767324c6d7a4ec2f07/extensions/theme-defaults/themes/dark_plus.json#L4). This approach could also be used for Pale Fire.

> This is a tool I have used to check if a set of colours meets WCAG 2.0 AA compliance

This looks very useful for ensuring that text can be read, but it doesn’t measure the ability to tell two colours apart. Correct me if I’m wrong, but is that not the issue you were having – telling modified files apart from untracked ones?

> But please don't take any of my criticisms above to heart, please keep iterating at your own pace, I'm just very happy to have these discussions and try to not be an unreasonable demander of other people's time. 😄

No, don’t worry about it. You’re actually giving helpful feedback that is useful to keep Pale Fire readable for everyone :)

**Comment by bbqsrc** [comment]
> This looks very useful for ensuring that text can be read, but it doesn’t measure the ability to tell two colours apart. Correct me if I’m wrong, but is that not the issue you were having – telling modified files apart from untracked ones?

Ah, sorry. I was trying to describe two related, but separate, issues: blocks of colour that are too similar next to each other (the first issue I raised), and the other issue was background colour lightness plus text lightness being hard to read.

Those colours in the side bar were warning and error colours.

As far as the first problem is concerned, I've only found it on the indicators in the scrollbar. The other issues were UI elements now having lighter backgrounds and much more muted text, which also strains my eyes. So yes, modified files vs untracked files is a bit of a strain. :smile:

I would note that any subtle changes to the actual syntax highlighting I haven't discerned yet, so everything seems perfectly fine there, just the UI chrome changes.

> I personally find it easy to read, but this is where themes become very subjective. I would prefer to avoid a situation in which anyone finds the theme so difficult to read they have to stick to an older version, so I think there are two choices:
>
>  * Add additional, higher-contrast colours to the palette and use them in the UI
>  * Add an option to revert to the old UI

Having "UI chrome" equivalent colours that respond to the environment they're in sounds like a very reasonable option, and I would note that creating a palette with slightly different values for the UI elements does not inherently mean creating a new palette, as colour perception is not 1:1 associated with the actual hex value for said colours due to how humans perceive colour.

I've been doing UI work on and off for quite a while now, and i can't remember a project where we didn't end up with a stylesheet of "colorname-brightnesslevel" nomenclature, like "blue-1", "blue-2", "blue-3" etc. :)

**Comment by matklad** [comment]
> project where we didn't end up with a stylesheet of "colorname-brightnesslevel" nomenclature, like "blue-1", "blue-2", "blue-3" etc. :)

https://github.com/matklad/pale-fire/blob/0504959768b5203a7934cfc4218d65ab2655fafc/themes/Pale%20Fire-color-theme.json#L390-L437

**Comment by bbqsrc** [comment]
Haha, ahhh, I even looked at this when it was linked earlier, but apparently I didn't even look at the names, just the colour values.

**Comment by arzg** [comment]
> As far as the first problem is concerned, I've only found it on the indicators in the scrollbar. The other issues were UI elements now having lighter backgrounds and much more muted text, which also strains my eyes. So yes, modified files vs untracked files is a bit of a strain. 😄

To me it seems like the biggest problems are the red and the yellow both being very unsaturated (pale?). I took a look at the screenshot of the terminal you sent and at first glance couldn’t tell the colour of `master` and`[$…` apart. Although I am not colourblind as far as I know (I took an online test once and it came up negative) and it is quite late where I am and I am pretty tired and my laptop screen is set to low brightness and I wasn’t really concentrating etc. etc. … I do see a big problem here. The image of the ‘overview ruler’ (scrollbar) you sent shows how much this is an issue even more, though: the yellow and the … err … red? Or is that orange?

> I would note that any subtle changes to the actual syntax highlighting I haven't discerned yet, so everything seems perfectly fine there, just the UI chrome changes.

Yup, that’s barely changed for Rust code.

> Having "UI chrome" equivalent colours that respond to the environment they're in sounds like a very reasonable option

With this do you mean the choice of making the UI higher-contrast? Or am I misunderstanding you here?

> and I would note that creating a palette with slightly different values for the UI elements does not inherently mean creating a new palette, as colour perception is not 1:1 associated with the actual hex value for said colours due to how humans perceive colour.

Sorry, I’m not really sure how human colour vision not being wholly correlated with literal RGB values really plays a role here, but you definitely do not want to get me started on that whole ‘perceptually uniform’ rabbit hole :)

**Comment by bbqsrc** [comment]
> With this do you mean the choice of making the UI higher-contrast? Or am I misunderstanding you here?

Yeah, I mean to say taking the base palette, dialing up the contrast and calling them the "UI" palette. :smile:

> Sorry, I’m not really sure how human colour vision not being wholly correlated with literal RGB values really plays a role here, but you definitely do not want to get me started on that whole ‘perceptually uniform’ rabbit hole :)

Haha, nah, I just meant that taking a dogmatic view of not changing colour values even if the perception of those colours is wrong due to surrounding colours is often unhelpful. Not suggesting you were taking any position on that, just waxing lyrical.  😆

**Comment by arzg** [comment]
> Yeah, I mean to say taking the base palette, dialing up the contrast and calling them the "UI" palette. 😄

Sounds like a plan! @matklad, what do you think would be the best approach here?

> Haha, nah, I just meant that taking a dogmatic view of not changing colour values even if the perception of those colours is wrong due to surrounding colours is often unhelpful.

Ah, I see what you were trying to say. I’m firmly in the camp of ‘perceptual uniform colour spaces and associated paraphernalia FTW’, so I think it’s definitely something that should be taken into account.

> waxing lyrical. 😆

Okay, now you’ve got me really confused. _Waxing lyrical?!_ I’ve never heard that term before! But don’t bother trying to explain, it’s better to keep the discussion on topic.

**Comment by arzg** [comment]
I think I’ll try creating a high-contrast version of Pale Fire using the new procedurally-generated colour palette.

**Comment by arzg** [comment]
I’ve pushed a high-contrast version with the name ‘Pale Fire High Contrast’. @bbqsrc would you mind trying it out? Here’s a comparison:

<img width="1513" alt="Screen Shot 2021-01-08 at 12 48 54 pm" src="https://user-images.githubusercontent.com/31783266/103964574-df280280-51af-11eb-8e86-569523054335.png">

<img width="1513" alt="Screen Shot 2021-01-08 at 12 49 08 pm" src="https://user-images.githubusercontent.com/31783266/103964586-e818d400-51af-11eb-8ed8-a0124a4b7fe4.png">

**Comment by bbqsrc** [comment]
```
Unable to load file:///Users/brendan/.vscode-insiders/extensions/matklad.pale-fire-0.3.11/themes/Pale%20Fire%20High%20Contrast-color-theme.json: Unable to read file '/Users/brendan/.vscode-insiders/extensions/matklad.pale-fire-0.3.11/themes/Pale Fire High Contrast-color-theme.json' (EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, open '/Users/brendan/.vscode-insiders/extensions/matklad.pale-fire-0.3.11/themes/Pale Fire High Contrast-color-theme.json')
```

It would seem the file is missing, as there's only 'Pale Fire-color-theme.json' at that path.

**Comment by bbqsrc** [comment]
Manually dropped the file in, is a significant improvement. :+1:

**Comment by arzg** [comment]
It should work normally since v0.3.12.

**Comment by arzg** [comment]
Do you think it would be a good idea to increase the contrast even further? It’s very easy to do now, so don’t hesitate to make suggestions :)

**Comment by arzg** [comment]
I found that the contrast between, for example, the editor area and the bar above it, lacking. I’ve pushed a commit that increases this contrast just for the high contrast theme. It also increases the differentiation between the various colour lightness levels used throughout the theme. For instance, the subtle difference between structs and primitive types should be slightly less subtle now :)
